File: notificare_geo/ios/Classes/NotificareGeoPluginBackgroundService.swift
//
//  NotificareGeoPluginBackgroundService.swift
//  notificare_geo
//
//  Created by Yevhenii Smirnov on 09/04/2024.
//

import NotificareKit
import NotificareGeoKit

fileprivate let DEFAULT_ERROR_CODE = "notificare_error"
fileprivate let NAMESPACE = "re.notifica.geo.flutter"

public class NotificareGeoPluginBackgroundService: NSObject, FlutterPlugin {
    internal static let instance = NotificareGeoPluginBackgroundService()
    
    private var registrar: FlutterPluginRegistrar?
    private var headlessRunner: FlutterEngine?
    private var callbackChannel: FlutterMethodChannel?
    
    private var isBackgroundStarting = false
    private var isInitialized = false
    private var pendingEvents: [BackgroundEvent] = []
    
    public static func register(with registrar: any FlutterPluginRegistrar) {
        instance.register(with: registrar)
    }
    
    private func register(with registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
        
        if (callbackChannel == nil) {
            headlessRunner = FlutterEngine(name: "NotificareGeoIsolate", project: nil, allowHeadlessExecution: true)
            callbackChannel = FlutterMethodChannel(
                name: "\(NAMESPACE)/notificare_geo_background",
                binaryMessenger: headlessRunner!.binaryMessenger,
                codec: FlutterJSONMethodCodec.sharedInstance()
            )
        }
    }
    
    private func startBackgroundService() {
        guard let callbackDispatcher = LocalStorage.getCallbackDispatcher() else { return }
        guard let callback = FlutterCallbackCache.lookupCallbackInformation(callbackDispatcher) else { return }
        
        let entrypoint = callback.callbackName
        let uri = callback.callbackLibraryPath
        
        if (headlessRunner?.run(withEntrypoint: entrypoint, libraryURI: uri) != true) {
            isBackgroundStarting = false
            return
        }
        
        registrar?.addMethodCallDelegate(self, channel: callbackChannel!)
        
        callbackChannel?.setMethodCallHandler { (call, result) in
            switch call.method {
            case "onBackgroundServiceInitialized": self.onBackgroundServiceInitialized(call, result)
                
                // Unhandled
            default: result(FlutterMethodNotImplemented)
            }
        }
    }
    
    private func onBackgroundServiceInitialized(_ call: FlutterMethodCall, _ response: FlutterResult) {
        processPendingEvents()
        isInitialized = true
        
        response(nil)
    }
}

// Background process
extension NotificareGeoPluginBackgroundService {
    private var applicationState: UIApplication.State {
        return UIApplication.shared.applicationState
    }
    
    internal func shouldProcessAsBackgroundEvent() -> Bool {
        return applicationState != .active && applicationState != .inactive
    }
    
    internal func processAsBackgroundEvent(event: BackgroundEvent) {
        if (event.callback == nil || LocalStorage.getCallbackDispatcher() == nil) { return }
        
        if (!isBackgroundStarting) {
            isBackgroundStarting = true
            startBackgroundService()
        }
        
        processEvent(event: event)
    }
    
    private func processEvent(event: BackgroundEvent) {
        if (!isInitialized) {
            pendingEvents.append(event)
            
            return
        }
        
        callbackChannel?.invokeMethod(event.method.rawValue, arguments: event.payload)
    }
    
    private func processPendingEvents() {
        while !pendingEvents.isEmpty {
            let event = pendingEvents.removeFirst()
            
            callbackChannel?.invokeMethod(event.method.rawValue, arguments: event.payload)
        }
    }
}

// Background Event
extension NotificareGeoPluginBackgroundService {
    internal enum Method: String, CaseIterable {
        case locationUpdated = "location_updated"
        case regionEntered = "region_entered"
        case regionExited = "region_exited"
        case beaconEntered = "beacon_entered"
        case beaconExited = "beacon_exited"
        case beaconsRanged = "beacons_ranged"
        case visit = "visit"
        case headingUpdated = "heading_updated"
    }
    
    internal enum CallbackType: String, CaseIterable {
        case locationUpdated = "re.notifica.geo.flutter.location_updated_callback"
        case regionEntered = "re.notifica.geo.flutter.region_entered_callback"
        case regionExited = "re.notifica.geo.flutter.region_exited_callback"
        case beaconEntered = "re.notifica.geo.flutter.beacon_entered_callback"
        case beaconExited = "re.notifica.geo.flutter.beacon_exited_callback"
        case beaconsRanged = "re.notifica.geo.flutter.beacons_ranged_callback"
        case visit = "re.notifica.geo.flutter.visit_callback"
        case headingUpdated = "re.notifica.geo.flutter.heading_updated_callback"
    }
    
    internal struct BackgroundEvent {
        let method: Method
        let callback: Int64?
        let payload: [String : Any?]
    }
}

extension NotificareGeoPluginBackgroundService {
    internal static func LocationUpdated(location: NotificareLocation) -> BackgroundEvent {
        let callback = LocalStorage.getCallback(callbackType: .locationUpdated)
        
        return BackgroundEvent(
            method: .locationUpdated,
            callback: callback,
            payload: [
                "callback" : callback,
                "location" : try! location.toJson()
            ]
        )
    }
    
    internal static func RegionEntered(region: NotificareRegion) -> BackgroundEvent {
        let callback = LocalStorage.getCallback(callbackType: .regionEntered)
        
        return BackgroundEvent(
            method: .regionEntered,
            callback: callback,
            payload: [
                "callback" : callback,
                "region" : try! region.toJson()
            ]
        )
    }
    
    internal static func RegionExited(region: NotificareRegion) -> BackgroundEvent {
        let callback = LocalStorage.getCallback(callbackType: .regionExited)
        
        return BackgroundEvent(
            method: .regionExited,
            callback: callback,
            payload: [
                "callback" : callback,
                "region" : try! region.toJson()
            ]
        )
    }
    
    internal static func BeaconEntered(beacon: NotificareBeacon) -> BackgroundEvent {
        let callback = LocalStorage.getCallback(callbackType: .beaconEntered)
        
        return BackgroundEvent(
            method: .beaconEntered,
            callback: callback,
            payload: [
                "callback" : callback,
                "beacon" : try! beacon.toJson()
            ]
        )
    }
    
    internal static func BeaconExited(beacon: NotificareBeacon) -> BackgroundEvent {
        let callback = LocalStorage.getCallback(callbackType: .beaconExited)
        
        return BackgroundEvent(
            method: .beaconExited,
            callback: callback,
            payload: [
                "callback" : callback,
                "beacon" : try! beacon.toJson()
            ]
        )
    }
    
    internal static func BeaconsRanged(beacons: [NotificareBeacon], region: NotificareRegion) -> BackgroundEvent {
        let callback = LocalStorage.getCallback(callbackType: .beaconsRanged)
        
        return BackgroundEvent(
            method: .beaconsRanged,
            callback: callback,
            payload: [
                "callback" : callback,
                "beacons" : try! beacons.map { try $0.toJson() },
                "region" : try! region.toJson()
            ]
        )
    }
    
    internal static func Visit(visit: NotificareVisit) -> BackgroundEvent {
        let callback = LocalStorage.getCallback(callbackType: .visit)
        
        return BackgroundEvent(
            method: .visit,
            callback: callback,
            payload: [
                "callback" : callback,
                "visit" : try! visit.toJson()
            ]
        )
    }
    
    internal static func HeadingUpdated(heading: NotificareHeading) -> BackgroundEvent {
        let callback = LocalStorage.getCallback(callbackType: .headingUpdated)
        
        return BackgroundEvent(
            method: .headingUpdated,
            callback: callback,
            payload: [
                "callback" : callback,
                "heading" : try! heading.toJson()
            ]
        )
    }
}

File: notificare_geo/android/src/main/kotlin/re/notifica/geo/flutter/NotificareGeoPluginBackgroundService.kt
package re.notifica.geo.flutter

import android.content.Context
import android.os.Handler
import android.util.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.embedding.engine.loader.FlutterLoader
import io.flutter.plugin.common.JSONMethodCodec
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.view.FlutterCallbackInformation
import re.notifica.geo.flutter.storage.NotificareGeoPluginStorage.Companion.getCallback
import re.notifica.geo.flutter.storage.NotificareGeoPluginStorage.Companion.getCallbackDispatcher
import re.notifica.geo.models.NotificareBeacon
import re.notifica.geo.models.NotificareLocation
import re.notifica.geo.models.NotificareRegion
import re.notifica.geo.models.toJson
import java.util.LinkedList
import java.util.Queue
import java.util.concurrent.atomic.AtomicBoolean

class NotificareGeoPluginBackgroundService : MethodCallHandler {
    private lateinit var backgroundChannel: MethodChannel
    private lateinit var backgroundHandler: Handler

    private var backgroundFlutterEngine: FlutterEngine? = null
    private val isServiceReady = AtomicBoolean(false)

    internal companion object {
        private const val NOTIFICARE_ERROR = "notificare_error"
        private val queue: Queue<BackgroundEvent> = LinkedList()
        internal val isAttachedToActivity = AtomicBoolean(false)

        @Volatile
        private var instance: NotificareGeoPluginBackgroundService? = null

        internal fun shouldProcessAsBackgroundEvent(): Boolean {
            synchronized(this) {
                if (isAttachedToActivity.get()) return false
            }

            return true
        }

        internal fun processAsBackgroundEvent(context: Context, event: BackgroundEvent) {
            if (event.callback == 0L) return

            synchronized(this) {
                if (instance == null) {
                    val callbackDispatcher = context.getCallbackDispatcher().also {
                        if (it == 0L) return
                    }

                    ensureInstanceInitialized(context, callbackDispatcher)
                }

                enqueueEvent(event)
            }
        }

        private fun ensureInstanceInitialized(context: Context, callbackDispatcher: Long) {
            instance = NotificareGeoPluginBackgroundService().also {
                it.startBackgroundService(context, callbackDispatcher)
            }
        }

        private fun enqueueEvent(event: BackgroundEvent) {
            instance?.also {
                it.handleBackgroundEvent(event)
                return
            }

            queue.add(event)
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "onBackgroundServiceInitialized" -> onBackgroundServiceInitialized()
            else -> result.notImplemented()
        }
    }

    private fun startBackgroundService(context: Context, callbackDispatcher: Long) {
        val flutterLoader = FlutterLoader()

        flutterLoader.startInitialization(context)
        flutterLoader.ensureInitializationComplete(context, null)

        if (backgroundFlutterEngine == null) {
            backgroundFlutterEngine = FlutterEngine(context)

            val callbackInfo = FlutterCallbackInformation.lookupCallbackInformation(callbackDispatcher)
            if (callbackInfo == null) {
                Log.e(NOTIFICARE_ERROR, "Failed to get callback information for a given handle.")
                return
            }

            val args = DartExecutor.DartCallback(
                context.assets,
                flutterLoader.findAppBundlePath(),
                callbackInfo
            )

            backgroundFlutterEngine!!.dartExecutor.executeDartCallback(args)
        }

        backgroundChannel = MethodChannel(
            backgroundFlutterEngine!!.dartExecutor.binaryMessenger,
            "re.notifica.geo.flutter/notificare_geo_background",
            JSONMethodCodec.INSTANCE
        )

        backgroundHandler = Handler(context.mainLooper)
        backgroundChannel.setMethodCallHandler(this)
    }

    private fun onBackgroundServiceInitialized() {
        synchronized(isServiceReady) {
            while (!queue.isEmpty()) {
                val event = queue.remove()
                processBackgroundEvent(event)
            }

            isServiceReady.set(true)
        }
    }

    private fun handleBackgroundEvent(event: BackgroundEvent) {
        synchronized(isServiceReady) {
            if (!isServiceReady.get()) {
                queue.add(event)
            } else {
                processBackgroundEvent(event)
            }
        }
    }

    private fun processBackgroundEvent(event: BackgroundEvent) {
        backgroundHandler.post {
            backgroundChannel.invokeMethod(event.method.id, event.payload)
        }
    }

    internal sealed class BackgroundEvent {
        abstract val method: Method
        abstract val callback: Long
        abstract val payload: Map<String, Any>

        internal enum class Method(val id: String) {
            LOCATION_UPDATED(id = "location_updated"),
            REGION_ENTERED(id = "region_entered"),
            REGION_EXITED(id = "region_exited"),
            BEACON_ENTERED(id = "beacon_entered"),
            BEACON_EXITED(id = "beacon_exited"),
            BEACONS_RANGED(id = "beacons_ranged"),
        }

        internal enum class CallbackType(val key: String) {
            LOCATION_UPDATED(key = "location_updated_callback"),
            REGION_ENTERED(key = "region_entered_callback"),
            REGION_EXITED(key = "region_exited_callback"),
            BEACON_ENTERED(key = "beacon_entered_callback"),
            BEACON_EXITED(key = "beacon_exited_callback"),
            BEACONS_RANGED(key = "beacons_ranged_callback"),
        }

        internal class LocationUpdated(
            context: Context,
            location: NotificareLocation
        ) : BackgroundEvent() {
            override val method = Method.LOCATION_UPDATED
            override val callback = context.getCallback(CallbackType.LOCATION_UPDATED)
            override val payload = mapOf(
                "callback" to callback,
                "location" to location.toJson()
            )
        }

        internal class RegionEntered(
            context: Context,
            region: NotificareRegion
        ) : BackgroundEvent() {
            override val method = Method.REGION_ENTERED
            override val callback = context.getCallback(CallbackType.REGION_ENTERED)
            override val payload = mapOf(
                "callback" to callback,
                "region" to region.toJson()
            )
        }

        internal class RegionExited(
            context: Context,
            region: NotificareRegion
        ) : BackgroundEvent() {
            override val method = Method.REGION_EXITED
            override val callback = context.getCallback(CallbackType.REGION_EXITED)
            override val payload = mapOf(
                "callback" to callback,
                "region" to region.toJson()
            )
        }

        internal class BeaconEntered(
            context: Context,
            beacon: NotificareBeacon
        ) : BackgroundEvent() {
            override val method = Method.BEACON_ENTERED
            override val callback = context.getCallback(CallbackType.BEACON_ENTERED)
            override val payload = mapOf(
                "callback" to callback,
                "beacon" to beacon.toJson()
            )
        }

        internal class BeaconExited(
            context: Context,
            beacon: NotificareBeacon
        ) : BackgroundEvent() {
            override val method = Method.BEACON_EXITED
            override val callback = context.getCallback(CallbackType.BEACON_EXITED)
            override val payload = mapOf(
                "callback" to callback,
                "beacon" to beacon.toJson()
            )
        }

        internal class BeaconsRanged(
            context: Context,
            beacons: List<NotificareBeacon>,
            region: NotificareRegion
        ) : BackgroundEvent() {
            override val method = Method.BEACONS_RANGED
            override val callback = context.getCallback(CallbackType.BEACONS_RANGED)
            override val payload = mapOf(
                "callback" to callback,
                "beacons" to beacons.map { it.toJson() },
                "region" to region.toJson()
            )
        }
    }
}
